quill/ecs/src/storage/blob_vec.rs

```diff
@@ -0,0 +1,289 @@
+use std::{
+    alloc::Layout,
+    mem::MaybeUninit,
+    ptr::{self, NonNull},
+    sync::Arc,
+};
+
+use crate::layout_ext::LayoutExt;
+use crate::space::MemorySpace;
+
+/// A vector of values of the same layout, stored
+/// as raw bytes.
+pub struct BlobVec {
+    /// Memory where the values are allocated.
+    space: Arc<MemorySpace>,
+    /// Layout of the values stored in the blob vector.
+    item_layout: Layout,
+
+    capacity: usize,
+    len: usize,
+    ptr: NonNull<u8>,
+}
+
+impl BlobVec {
+    pub fn new(item_layout: Layout) -> Self {
+        Self::with_space(item_layout, Arc::new(MemorySpace::host()))
+    }
+
+    pub fn with_space(item_layout: Layout, space: Arc<MemorySpace>) -> Self {
+        // Create a dangling pointer.
+        let ptr = NonNull::new(item_layout.align() as *mut u8).expect("align > 0");
+
+        Self {
+            space,
+            item_layout,
+
+            capacity: 0,
+            len: 0,
+            ptr,
+        }
+    }
+
+    pub fn push<T>(&mut self, value: T) {
+        self.assert_layout_matches::<T>();
+
+        let value = MaybeUninit::new(value);
+        unsafe {
+            self.push_raw(value.as_ptr().cast());
+        }
+    }
+
+    pub unsafe fn push_raw(&mut self, value: *const u8) {
+        let new_len = self.len.checked_add(1).expect("length overflow");
+        if new_len > self.capacity {
+            self.grow();
+            debug_assert!(new_len <= self.capacity);
+        }
+
+        self.len = new_len;
+
+        unsafe {
+            ptr::copy_nonoverlapping(value, self.item_ptr(self.len - 1), self.item_layout.size());
+        }
+
+        self.check_invariants();
+    }
+
+    fn grow(&mut self) {
+        let old_capacity = self.capacity;
+        let new_capacity = match old_capacity {
+            0 => 4,
+            x => x.checked_mul(2).expect("capacity overflow"),
```

Comment:
`next_power_of_two` could maybe be a good alternative?

Reply:
I would assume the `next_power_of_two` function is more complex than a simple multiplication and I doubt the compiler can elide that extra logic because it can't make assumptions about the values of `self`. What would be the benefit?

Reply:
I'm not saying there is anything wrong with this implementation, I'm only suggesting `next_power_of_two` since that is how `RawVec`, by extension `Vec`, is implemented.

Reply:
Actually from what I can see the next_power_of_two function just does bit shifting.

Reply:
I cannot find a call to `next_power_of_two` anywhere in `raw_vec.rs`...
`grow_amortized` uses a simple `* 2` multiplication.
Also, looking at the implementation `next_power_of_two` is noticeably more complex than a simple multiplication.

Reply:
I've been lied to.